### PR TITLE
fix($parse): call once stable bind-once expressions with filter

### LINKED ISSUE #14583: in a one-time filtered expression (:: func() | filter), the filter is invoked the second time even though its input has not changed
https://plnkr.co/edit/S9VhW6TdvkvtpGhgQBKQ?p=preview

for a regular expression the filter is invoked only once

### PR DESCRIPTION
**Pull request by lgalfaso**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
fix, but given the nature I would submit this for 1.6 only

**What is the current behavior? (You can also link to an open issue here)**
stateless filters on bind-once expression are called multiple times.

**What is the new behavior (if this is a feature change)?**
stateless filters on bind-once expression are called once.

**Does this PR introduce a breaking change?**
No, but developers might expect the multiple calls (even when they should not)

**Please check if the PR fulfills these requirements**
- [X] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [X] Tests for the changes have been added (for bug fixes / features)

**Other information**:

When an expression has the following:
- Bind-once
- A stateless filter at the end
- Optionally an expression interceptor

Then call the filter in the expression once.

Closes: #14583

### CODE REVIEW

File: src/ng/parse.js [hunk 1]
```diff
@@ -2050,14 +2050,20 @@ function $ParseProvider() {
...
+      }
+      function oneTimeWatch(scope) {
         return parsedExpression(scope);
-      }, function oneTimeListener(value, old, scope) {
+      }
+      function oneTimeListener(value, old, scope) {
         lastValue = value;
         if (isFunction(listener)) {
-          listener.apply(this, arguments);
+          listener(value, old, scope);
```
Comment (jbedard) [inline src/ng/parse.js]: This (and the removal of `.call` below) seem to be unrelated to the rest of the PR?

I think the change is ok since it is equivalent to the [normal listener invoking](https://github.com/angular/angular.js/blob/4487d4a8a477ad440500e3574c141da48d13ab3d/src/ng/rootScope.js#L803). But I think the idea of doing .call/apply here was to avoid duplicating that logic.

Comment (lgalfaso) [inline src/ng/parse.js]: `this` is always undefined, so the old form adds nothing.

Comment (jbedard) [inline src/ng/parse.js]: The old form just preserved whatever the normal listener invoking did, so the logic (or lack of) was not duplicated. I'm fine with the change, but it seems unrelated to the actual fix?

File: src/ng/parse.js [hunk 2]
```diff
@@ -2050,14 +2050,20 @@ function $ParseProvider() {
...
     }
 
-    function oneTimeWatchDelegate(scope, listener, objectEquality, parsedExpression) {
+    function oneTimeWatchDelegate(scope, listener, objectEquality, parsedExpression, prettyPrintExpression) {
       var unwatch, lastValue;
-      return unwatch = scope.$watch(function oneTimeWatch(scope) {
+      if (parsedExpression.inputs) {
+        return unwatch = inputsWatchDelegate(scope, oneTimeListener, objectEquality, parsedExpression, prettyPrintExpression);
+      } else {
+        return unwatch = scope.$watch(oneTimeWatch, oneTimeListener, objectEquality);
```
Comment (gkalpak) [inline src/ng/parse.js]: Why not use `parsedExpression` directly ?

File: test/ng/parseSpec.js [hunk 3]
```diff
@@ -3232,6 +3232,18 @@ describe('parser', function() {
           expect(fn()).toEqual(undefined);
         }));
 
+        it('should invoke a stateless filter once when the parsed expression has an interceptor',
+           inject(function($parse, $rootScope) {
+          var countFilter = jasmine.createSpy();
+          var interceptor = jasmine.createSpy();
+          countFilter.and.callThrough();
```
Comment (gkalpak) [inline test/ng/parseSpec.js]: Is this necessary ?

File: src/ng/parse.js (hunk 2 above)
Comment (lgalfaso) [inline src/ng/parse.js]: `parsedExpression` has `$$watchDelegate` :)

File: test/ng/parseSpec.js (hunk 3 above)
Comment (lgalfaso) [inline test/ng/parseSpec.js]: In a strict sense, no, but there is some special logic for one time binding and when something is evaluated to `undefined` and I want to be sure that whoever reads the test, that we are not running into that case.

Comment (gkalpak) [inline test/ng/parseSpec.js]: But, how is `.and.callThrough()` changing anything. You don't have an actual function to call through anyway.

Comment (lgalfaso) [inline test/ng/parseSpec.js]: I do not understand this. `countFilter` is registered as a filter and then this filter is used in the expression `foo() | count`. `foo` is a function that returns `1` and i want to be sure that the expression evaluates to something other than `undefined`

Comment (gkalpak) [inline test/ng/parseSpec.js]: `countFilter` is initialized as `jasmine.createSpy()` (2 lines above). So you are basically doing `jasmine.createSpy().and.callThrough()`. What you are calling through to is `function() {}` which should return undefined no matter what (e.g. see [demo][1]).

[1]: https://codepen.io/ExpertSystem/pen/LNoPqe?editors=0010

Comment (lgalfaso) [inline test/ng/parseSpec.js]: you are right, my mistake. Fixed.

**Comment by gkalpak** [comment]
Should we backport this to 1.5.x (since it's a fix) ?

**Comment by lgalfaso** [comment]
> Should we backport this to 1.5.x (since it's a fix) ?

I lean towards not to backport this. The reason being that it does add some behavior change.
